components/capi/src/pluralrules.rs

```diff
@@ -0,0 +1,129 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use icu_locid::Locale as ICULocale;
+use icu_plurals::PluralCategory as ICUPluralCategory;
+use icu_plurals::PluralOperands as ICUPluralOperands;
+use icu_plurals::PluralRuleType as ICUPluralRuleType;
+
+use crate::provider::ErasedDataProvider;
+
+use std::ptr;
+
+// opaque type
+pub use icu_plurals::PluralRules;
+
+#[repr(C)]
+pub struct CreatePluralRulesResult {
+    rules: *mut PluralRules,
+    success: bool,
+}
+
+#[no_mangle]
+pub extern "C" fn plural_rules_create(
+    locale: &ICULocale,
+    provider: &mut ErasedDataProvider,
+    ty: PluralRuleType,
```

Comment:
This is an example of where we should think about the implications for data slicing.  There are two data keys for PluralRules: CARDINAL and ORDINAL.  In Rust, you can pass in the type, and code slicing will remove the unused data key for you.  However, in C, you get the whole function, so both data keys are compiled, even if you need only one.

We could:

1. Manually make custom entrypoints for each Rust API that have one function per data key, like `icu4x_plural_rules_open_cardinal()` and `icu4x_plural_rules_open_ordinal()`
2. Do something more along the lines of what I suggested in wrapper_layer.md, where the C functions wrap Rust one layer below the ergonomic Rust API.
3. Ignore the problem 😢.
3. Other?

Reply:
I think (1.) is fine. I wouldn't include it in the API just yet but we can as needed I think.

A problem will be if we wish to support non-erased data providers, we'll have multiple `_create` functions from the angle of the data providers. So I'm wary of a multiplicative blowup in constructors.

Reply:
I was originally thinking that we would do something more along the lines of:

- The entrypoint APIs like `icu4x_plural_rules_create()` take a data struct, not a data provider
- The code to resolve from data provider to data struct is written in the wrapper layer

Since the data provider may be async, this approach is cleaner since it doesn't require that `icu4x_plural_rules_create()` also be async.

Reply:
How about if we did something along the lines of

```cpp
ICU4X_Provider_PluralRulesV1* data = icu4x_provider_load_plural_rules_cardinal(provider, ...);
ICU4X_PluralRules* rules = icu4x_plural_rules_create(data, ...);
```

Reply:
Hmmm... this does make it easier for async and stuff.

Do we have any cases where the ICU4X object (like PluralRules) would need to care about the data provider _after_ it has been constructed?

I think we should land the current code as is and file a followup for improving the data provider FFI story, because first off we'd have to add internal APIs for this everywhere.